drivers: Perform a runtime check if a driver is capable of an operation

lpereira opened the pull request:
Driver APIs might not implement all operations, making it possible for a user thread to get the kernel to execute a function at 0x00000000.

Perform runtime checks in all the driver handlers, checking if they're capable of performing the requested operation.

Fixes #6907.

(Checkpatch will whine on this patch with an error due to the macro in uart_handlers.  The suggestion (putting parenthesis around macro arguments) does not work in this case.)

lpereira committed:
drivers: Perform a runtime check if a driver is capable of an operation

Driver APIs might not implement all operations, making it possible for
a user thread to get the kernel to execute a function at 0x00000000.

Perform runtime checks in all the driver handlers, checking if they're
capable of performing the requested operation.

Fixes #6907.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

dbkinder commented on doc/kernel/usermode/syscalls.rst:
```diff
@@ -292,6 +292,9 @@ Several macros exist to validate arguments:
   a message parameter, instead printing the expression tested if it
   fails. The latter should only be used for the most obvious of tests.
 
+* :c:macro:`_SYSCALL_DRIVER_OP()` checks in runtime if a driver
```
``checks at runtime`` sounds better

lpereira replied:
Done

dbkinder reviewed (approved):
thanks!

nashif commented on drivers/adc/adc_handlers.c:
```diff
@@ -7,8 +7,21 @@
 
-_SYSCALL_HANDLER1_SIMPLE_VOID(adc_enable, K_OBJ_DRIVER_ADC, struct device *);
-_SYSCALL_HANDLER1_SIMPLE_VOID(adc_disable, K_OBJ_DRIVER_ADC, struct device *);
+_SYSCALL_HANDLER(adc_enable, dev)
```
well, those APIs are going away AFAIK, there is a new ADC API being proposed....

andrewboie replied:
strongly concur with @lpereira

andrewboie commented on drivers/adc/adc_handlers.c:
```diff
@@ -7,8 +7,21 @@
 #include <adc.h>
 #include <syscall_handler.h>
 
-_SYSCALL_HANDLER1_SIMPLE_VOID(adc_enable, K_OBJ_DRIVER_ADC, struct device *);
```
I hate to see all these oneliners go away.
Consider _SYSCALL_DRIVER_HANDLER1_SIMPLE_VOID(api_name, op_name, api_struct, ...)
Not mandatory

lpereira committed:
scripts: gen_kobject_list: Generate enums and case statements

Adding a new kernel object type or driver subsystem requires changes
in various different places.  This patch makes it easier to create
those devices by generating as much as possible in compile time.

No behavior change.

Signed-off-by: Leandro Pereira <leandro.pereira@intel.com>

SebastianBoe commented on scripts/gen_kobject_list.py:
```diff
@@ -108,18 +108,94 @@ def write_gperf_table(fp, eh, objs, static_begin, static_end):
+		_SYSCALL_OBJ(ptr, K_OBJ_DRIVER_%(type)s); \\
+		_SYSCALL_DRIVER_OP(ptr, %(api_struct)s, op); \\
+	} while (0)
+"""
```
It is easier for users of generated code to understand it when it is human-readable.

Could you look into refactoring away some of the redundancy,
e.g. something like this:

```
#define _SYSCALL_DRIVER_GEN(ptr, op, driver_lower_case, driver_upper_case) \
	do {                                             \
		_SYSCALL_OBJ(ptr, K_OBJ_DRIVER_##driver_upper_case);    \
		_SYSCALL_DRIVER_OP(ptr, driver_lower_case##_driver_api, op); \
	} while (0)

#define _SYSCALL_DRIVER_ADC(ptr, op)     _SYSCALL_DRIVER_GEN(ptr, op, adc, ADC)
#define _SYSCALL_DRIVER_AIO_CMP(ptr, op) _SYSCALL_DRIVER_GEN(ptr, op, aio_cmp, AIO_CMP)


```

Instead of:

```
#define _SYSCALL_DRIVER_COUNTER(ptr, op) \
	do { \
		_SYSCALL_OBJ(ptr, K_OBJ_DRIVER_COUNTER); \
		_SYSCALL_DRIVER_OP(ptr, counter_driver_api, op); \
	} while (0)

#define _SYSCALL_DRIVER_CRYPTO(ptr, op) \
	do { \
		_SYSCALL_OBJ(ptr, K_OBJ_DRIVER_CRYPTO); \
		_SYSCALL_DRIVER_OP(ptr, crypto_driver_api, op); \
	} while (0)


```

Would make it clear at a glance that all _SYSCALL_DRIVER_* calls consist of a _OBJ call and a DRIVER_OP call.

(This also ties into the general principle that you want to do as much as possible in the source language, and as little as possible in the build system / codegen scripts).

lpereira replied:
This is a kind of po-tah-to/po-tay-to situation, but, sure, let me make the change.

andrewboie commented on drivers/i2c/i2c_handlers.c:
```diff
@@ -9,7 +9,7 @@
 _SYSCALL_HANDLER(i2c_configure, dev, dev_config)
 {
-	_SYSCALL_OBJ(dev, K_OBJ_DRIVER_I2C);
+	_SYSCALL_DRIVER_I2c(dev, configure);
```
_SYSCALL_DRIVER_I2c not in all caps, build error

tyrantlizard77 reviewed (changes requested):
there's some build problems revealed by CI.
one is a typo. on some unit test cases the header isn't being generated, but I'm not sure why.

andrewboie commented:
`/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/include/kernel.h:142: warning: include file kobj-types-enum.h not found, perhaps you forgot to add its directory to INCLUDE_PATH?
`

Documentation build is breaking

dbkinder commented:
Where is ``kobj-types-enum.h`` to be found?

dbkinder commented on include/kernel.h:
```diff
@@ -135,35 +135,11 @@ struct k_mem_partition;
+	 * gen_kobject_list.py.  It includes basic kernel objects (e.g.  pipes
+	 * and mutexes) and driver types.
+	 */
+#include <kobj-types-enum.h>
```
Doxygen is looking for this include file when generating API documentation.  Either we need an empty file as a place holder in the folder structure (maybe with just a comment saying how it will be generated), or we need to tell doxygen to just ignore this include with:

    /** @cond 
     *  Doxygen should ignore this build-time generated include file when
     *  generating API documentation
     */

    #include <kobj-types-enum.h>

    /** @endcond
     */

lpereira commented:
Rebased with Doxygen fix.

dbkinder reviewed (approved):
doxygen fix solved the doc build problem, yea, but now checkpatch isn't happy...

lpereira commented:
The checkpatch warnings are false positives.

nashif commented:
rebase

lpereira commented:
Rebased.

On 04/24/2018 06:50 PM, Anas Nashif wrote:
> rebase
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/zephyrproject-rtos/zephyr/pull/6930#issuecomment-384135501>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAA6mQE4ErKDh7fclhl-H8Fb3Xfhj50Uks5tr9ZZgaJpZM4THe_I>.
>
